File: niko-studio/src/workflow/revision_loop.py
# -*- coding: utf-8 -*-
"""
修订循环模块 (Revision Loop)

实现 Critic 反馈驱动的 Writer 修订循环。
支持最大循环次数限制和人工介入触发。
"""

from typing import Dict, Any, Optional, Callable, Awaitable
from dataclasses import dataclass, field
from enum import Enum


class RevisionDecision(Enum):
    """修订决策"""
    APPROVED = "APPROVED"           # 通过，无需修订
    REVISE = "REVISE"               # 需要修订
    REWRITE = "REWRITE"             # 需要重写
    HUMAN_REVIEW = "HUMAN_REVIEW"   # 需要人工审阅


@dataclass
class RevisionConfig:
    """修订循环配置"""
    max_revisions: int = 3          # 最大修订次数
    pass_score: float = 80.0        # 通过分数阈值
    min_c_score: float = 7.0        # C(冲突)维度最低分
    human_review_score: float = 70.0  # 触发人工审阅的分数
    score_improvement_threshold: float = 5.0  # 最小分数提升阈值


@dataclass
class RevisionState:
    """修订状态"""
    revision_count: int = 0
    current_score: float = 0.0
    previous_score: float = 0.0
    decision: RevisionDecision = RevisionDecision.REVISE
    feedback: str = ""
    history: list = field(default_factory=list)
    stagnant_count: int = 0  # 分数停滞次数


class RevisionLoop:
    """
    修订循环控制器

    管理 Writer ↔ Critic 的修订循环，包括：
    - 分数追踪和趋势分析
    - 停滞检测（连续无提升）
    - 自动人工介入触发
    """

    def __init__(self, config: Optional[RevisionConfig] = None):
        self.config = config or RevisionConfig()
        self.state = RevisionState()

    def should_continue(self) -> bool:
        """判断是否应该继续修订循环"""
        # 已通过
        if self.state.decision == RevisionDecision.APPROVED:
            return False

        # 需要人工审阅
        if self.state.decision == RevisionDecision.HUMAN_REVIEW:
            return False

        # 达到最大修订次数
        if self.state.revision_count >= self.config.max_revisions:
            return False

        # 分数停滞超过2次
        if self.state.stagnant_count >= 2:
            return False

        return True

    def update_from_critic(self, critic_result: Dict[str, Any]) -> RevisionDecision:
        """
        根据 Critic 结果更新状态

        Args:
            critic_result: Critic Agent 的输出

        Returns:
            RevisionDecision: 修订决策
        """
        self.state.previous_score = self.state.current_score
        self.state.current_score = critic_result.get("total_score", 0)
        self.state.revision_count += 1

        # 获取原始决策
        raw_decision = critic_result.get("decision", "REVISE")
        self.state.feedback = critic_result.get("actionable_feedback", "")

        # 记录历史
        self.state.history.append({
            "revision": self.state.revision_count,
            "score": self.state.current_score,
            "decision": raw_decision,
            "feedback_preview": self.state.feedback[:100] if self.state.feedback else ""
        })

        # 检查分数提升
        score_improvement = self.state.current_score - self.state.previous_score
        if self.state.revision_count > 1 and score_improvement < self.config.score_improvement_threshold:
            self.state.stagnant_count += 1
        else:
            self.state.stagnant_count = 0

        # 确定最终决策
        decision = self._determine_decision(raw_decision, critic_result)
        self.state.decision = decision

        return decision

    def _determine_decision(
        self,
        raw_decision: str,
        critic_result: Dict[str, Any]
    ) -> RevisionDecision:
        """确定最终修订决策"""
        score = self.state.current_score

        # 检查 LOCK 分析中的 C 分数
        lock_analysis = critic_result.get("lock_analysis", {})
        c_score = 10  # 默认满分
        if lock_analysis:
            c_data = lock_analysis.get("C", {})
            if isinstance(c_data, dict):
                c_score = c_data.get("score", 10)

        # 决策逻辑
        if raw_decision == "APPROVED":
            if score >= self.config.pass_score and c_score >= self.config.min_c_score:
                return RevisionDecision.APPROVED
            else:
                # 分数不够，降级为修订
                return RevisionDecision.REVISE

        if raw_decision == "REWRITE":
            if self.state.revision_count >= self.config.max_revisions:
                return RevisionDecision.HUMAN_REVIEW
            return RevisionDecision.REWRITE

        if raw_decision == "HUMAN_REVIEW":
            return RevisionDecision.HUMAN_REVIEW

        # REVISE 情况
        if self.state.revision_count >= self.config.max_revisions:
            return RevisionDecision.HUMAN_REVIEW

        if self.state.stagnant_count >= 2:
            # 分数停滞，需要人工介入
            return RevisionDecision.HUMAN_REVIEW

        return RevisionDecision.REVISE

    def get_feedback_for_writer(self) -> Dict[str, Any]:
        """
        获取传递给 Writer 的反馈信息

        Returns:
            包含 issues, suggestions, dimension_scores 的字典
        """
        return {
            "feedback": self.state.feedback,
            "revision_count": self.state.revision_count,
            "previous_score": self.state.previous_score,
            "current_score": self.state.current_score,
            "history": self.state.history,
        }

    def get_summary(self) -> Dict[str, Any]:
        """获取修订循环摘要"""
        return {
            "total_revisions": self.state.revision_count,
            "final_score": self.state.current_score,
            "final_decision": self.state.decision.value,
            "score_trend": [h["score"] for h in self.state.history],
            "stagnant_count": self.state.stagnant_count,
            "history": self.state.history,
        }

    def reset(self):
        """重置修订状态"""
        self.state = RevisionState()


async def run_revision_loop(
    draft: str,
    scene_card: Dict[str, Any],
    writer_fn: Callable[[str, Dict[str, Any]], Awaitable[str]],
    critic_fn: Callable[[str, Dict[str, Any]], Awaitable[Dict[str, Any]]],
    config: Optional[RevisionConfig] = None,
    verbose: bool = True
) -> Dict[str, Any]:
    """
    执行完整的修订循环

    Args:
        draft: 初始草稿
        scene_card: 场景卡片
        writer_fn: Writer 修订函数 (draft, feedback) -> revised_draft
        critic_fn: Critic 评估函数 (draft, scene_card) -> critic_result
        config: 修订配置
        verbose: 是否输出日志

    Returns:
        包含最终草稿和修订历史的字典
    """
    loop = RevisionLoop(config)
    current_draft = draft

    while True:
        # 评估当前草稿
        if verbose:
            print(f"\n🧐 第 {loop.state.revision_count + 1} 次评估...")

        critic_result = await critic_fn(current_draft, scene_card)
        decision = loop.update_from_critic(critic_result)

        if verbose:
            print(f"   分数: {loop.state.current_score:.1f}")
            print(f"   决策: {decision.value}")

        # 检查是否应该继续
        if not loop.should_continue():
            break

        # 获取反馈并修订
        feedback = loop.get_feedback_for_writer()

        if verbose:
            print(f"\n✍️ 第 {loop.state.revision_count} 次修订...")

        current_draft = await writer_fn(current_draft, feedback)

    # 返回结果
    summary = loop.get_summary()
    summary["final_draft"] = current_draft

    if verbose:
        print(f"\n{'='*50}")
        print(f"修订循环完成")
        print(f"   总修订次数: {summary['total_revisions']}")
        print(f"   最终分数: {summary['final_score']:.1f}")
        print(f"   最终决策: {summary['final_decision']}")
        print(f"{'='*50}")

    return summary

File: src-ts/workflow/revision-loop.ts
/**
 * Revision Loop Module
 *
 * Implements Critic feedback-driven Writer revision loop.
 * Supports max-loop limit and human-intervention triggers.
 *
 * Migrated from src/workflow/revision_loop.py
 */

import type { QualityMode, QualityLevel } from './novel-state.js';
import {
  NOVEL_PASS_SCORE,
  NOVEL_MIN_C_SCORE,
  NOVEL_HUMAN_REVIEW_SCORE,
  NOVEL_SCORE_IMPROVEMENT_THRESHOLD,
} from './novel-state.js';
import { type ContentType as SessionContentType } from './session/session-manager.js';
import { createLogger } from '../logger/index.js';

const _log = createLogger('workflow-revision-loop');

// ============================================================
// Types
// ============================================================

export const QUALITY_LEVEL_ORDER: readonly QualityLevel[] = ['ultra', 'high', 'medium', 'fluent'] as const;

export enum RevisionDecision {
  APPROVED = 'APPROVED',
  REVISE = 'REVISE',
  REWRITE = 'REWRITE',
  HUMAN_REVIEW = 'HUMAN_REVIEW',
}

// ============================================================
// RevisionConfig
// ============================================================

export interface RevisionConfig {
  max_revisions: number;
  pass_score: number;
  min_c_score: number;
  human_review_score: number;
  score_improvement_threshold: number;
  quality_mode: QualityMode;
  quality_level: QualityLevel;
  degrade_on_timeout: boolean;
  degrade_on_error: boolean;
  quality_phase_timeout_seconds: number;
}

export const DEFAULT_REVISION_CONFIG: RevisionConfig = {
  max_revisions: 3,
  pass_score: NOVEL_PASS_SCORE,
  min_c_score: NOVEL_MIN_C_SCORE,
  human_review_score: NOVEL_HUMAN_REVIEW_SCORE,
  score_improvement_threshold: NOVEL_SCORE_IMPROVEMENT_THRESHOLD,
  quality_mode: 'auto',
  quality_level: 'high',
  degrade_on_timeout: true,
  degrade_on_error: true,
  quality_phase_timeout_seconds: 30,
};

// ============================================================
// RevisionState
// ============================================================

export interface RevisionState {
  revision_count: number;
  current_score: number;
  previous_score: number;
  decision: RevisionDecision;
  feedback: string;
  history: Array<Record<string, unknown>>;
  stagnant_count: number;
  checkpoint_trace: Array<Record<string, unknown>>;
  last_checkpoint_id: string;
  quality_mode: QualityMode;
  requested_quality_level: QualityLevel;
  effective_quality_level: QualityLevel;
  degrade_reason: string;
  degrade_steps: Array<Record<string, unknown>>;
  feedback_artifacts: Array<Record<string, unknown>>;
}

function createRevisionState(
  qualityMode: QualityMode,
  requestedLevel: QualityLevel,
  effectiveLevel: QualityLevel,
): RevisionState {
  return {
    revision_count: 0,
    current_score: 0.0,
    previous_score: 0.0,
    decision: RevisionDecision.REVISE,
    feedback: '',
    history: [],
    stagnant_count: 0,
    checkpoint_trace: [],
    last_checkpoint_id: '',
    quality_mode: qualityMode,
    requested_quality_level: requestedLevel,
    effective_quality_level: effectiveLevel,
    degrade_reason: '',
    degrade_steps: [],
    feedback_artifacts: [],
  };
}

// ============================================================
// RevisionLoop class
// ============================================================

export class RevisionLoop {
  public config: RevisionConfig;
  public state: RevisionState;
  public checkpointStore: Record<string, unknown>;

  constructor(config?: Partial<RevisionConfig>, checkpointStore?: Record<string, unknown>) {
    this.config = { ...DEFAULT_REVISION_CONFIG, ...config };
    const normalizedLevel = RevisionLoop.normalizeQualityLevel(this.config.quality_level);
    this.state = createRevisionState(
      this.config.quality_mode,
      normalizedLevel,
      normalizedLevel,
    );
    this.checkpointStore = checkpointStore ?? {};
  }

  static normalizeQualityLevel(level: string): QualityLevel {
    if ((QUALITY_LEVEL_ORDER as readonly string[]).includes(level)) return level as QualityLevel;
    return 'high';
  }

  static nextQualityLevel(level: QualityLevel): QualityLevel {
    const index = QUALITY_LEVEL_ORDER.indexOf(level);
    if (index >= QUALITY_LEVEL_ORDER.length - 1) return level;
    return QUALITY_LEVEL_ORDER[index + 1];
  }

  private _recordDegradeStep(reason: string, phase: string): boolean {
    if (this.state.quality_mode !== 'auto') return false;

    const fromLevel = this.state.effective_quality_level;
    const toLevel = RevisionLoop.nextQualityLevel(fromLevel);
    if (toLevel === fromLevel) return false;

    this.state.effective_quality_level = toLevel;
    this.state.degrade_reason = reason;
    this.state.degrade_steps.push({
      from_level: fromLevel,
      to_level: toLevel,
      reason,
      phase,
      timestamp: new Date().toISOString(),
    });
    return true;
  }

  handleRuntimeEvent(
    event: 'timeout' | 'error',
    phase: string,
    detail: string = '',
  ): boolean {
    if (event === 'timeout' && !this.config.degrade_on_timeout) return false;
    if (event === 'error' && !this.config.degrade_on_error) return false;

    let reason = `${event}:${phase}`;
    if (detail) reason = `${reason}:${detail}`;
    return this._recordDegradeStep(reason, phase);
  }

  shouldContinue(): boolean {
    if (this.state.decision === RevisionDecision.APPROVED) return false;
    if (this.state.decision === RevisionDecision.HUMAN_REVIEW) return false;
    if (this.state.revision_count >= this.config.max_revisions) return false;
    if (this.state.stagnant_count >= 2) return false;
    return true;
  }

  // -----------------------------------------------
  // Checkpoint helpers
  // -----------------------------------------------

  private _buildRoundIdentifier(): string {
    return `round-${this.state.revision_count}`;
  }

  private _buildCheckpointArtifact(
    roundIdentifier: string,
    criticResult: Record<string, unknown>,
  ): Record<string, unknown> {
    return {
      checkpoint_id: `revision-${roundIdentifier}`,
      revision_count: this.state.revision_count,
      round_identifier: roundIdentifier,
      step_id: roundIdentifier,
      stage: 'critic',
      decision: this.state.decision,
      score: this.state.current_score,
      previous_score: this.state.previous_score,
      stagnant_count: this.state.stagnant_count,
      session_id: String(criticResult['session_id'] ?? ''),
      trace: {
        state_trace_id: `${roundIdentifier}:${this.state.current_score}`,
        revision_checkpoint_id: `revision-${roundIdentifier}`,
      },
    };
  }

  private _persistCheckpointArtifact(artifact: Record<string, unknown>): string {
    const checkpointId = String(artifact['checkpoint_id'] ?? '');
    if (!checkpointId) return '';

    this.checkpointStore[checkpointId] = artifact;
    this.state.last_checkpoint_id = checkpointId;
    this.state.checkpoint_trace.push({
      checkpoint_id: checkpointId,
      step_id: String(artifact['step_id'] ?? ''),
      stage: String(artifact['stage'] ?? 'critic'),
      round_identifier: String(artifact['round_identifier'] ?? ''),
    });

    return checkpointId;
  }

  // -----------------------------------------------
  // Feedback artifact helpers
  // -----------------------------------------------

  private _normalizeSeverity(priority: string): string {
    const normalized = (priority ?? '').trim().toLowerCase();
    if (normalized === 'high' || normalized === 'critical') return 'high';
    if (normalized === 'medium' || normalized === 'med') return 'medium';
    if (normalized === 'low' || normalized === 'minor') return 'low';
    return 'medium';
  }

  private _inferScope(anchor: string): string {
    const token = (anchor ?? '').toLowerCase();
    if (token.includes('scene') || token.includes('\u573a\u666f')) return 'scene';
    return 'chapter';
  }

  private _buildFeedbackArtifacts(
    roundIdentifier: string,
    criticResult: Record<string, unknown>,
  ): Array<Record<string, unknown>> {
    const instructions = criticResult['revision_instructions'];
    const actionableFeedback = String(criticResult['actionable_feedback'] ?? '').trim();

    const artifacts: Array<Record<string, unknown>> = [];

    if (Array.isArray(instructions) && instructions.length > 0) {
      instructions.forEach((item: unknown, index: number) => {
        if (typeof item !== 'object' || item === null) return;
        const entry = item as Record<string, unknown>;

        const anchor = String(entry['target'] ?? '').trim() || `chapter-${this.state.revision_count}`;
        const issue = String(entry['issue'] ?? '').trim() || 'unspecified issue';
        const recommendation = String(entry['suggestion'] ?? '').trim() || actionableFeedback || 'revise content';
        const priority = String(entry['priority'] ?? 'medium');
        const severity = this._normalizeSeverity(priority);

        artifacts.push({
          feedback_id: `feedback-${roundIdentifier}-${index + 1}`,
          round_id: roundIdentifier,
          scope: this._inferScope(anchor),
          anchor,
          severity,
          issue,
          recommendation,
          source: 'critic',
        });
      });
    }

    if (artifacts.length === 0) {
      const defaultAnchor = `chapter-${this.state.revision_count}`;
      const defaultIssue = actionableFeedback || 'quality improvement needed';
      artifacts.push({
        feedback_id: `feedback-${roundIdentifier}-1`,
        round_id: roundIdentifier,
        scope: 'chapter',
        anchor: defaultAnchor,
        severity: 'medium',
        issue: defaultIssue,
        recommendation: actionableFeedback || 'revise according to critic feedback',
        source: 'critic',
      });
    }

    return artifacts;
  }

  // -----------------------------------------------
  // Core update logic
  // -----------------------------------------------

  updateFromCritic(criticResult: Record<string, unknown>): RevisionDecision {
    this.state.previous_score = this.state.current_score;
    this.state.current_score = typeof criticResult['total_score'] === 'number'
      ? criticResult['total_score'] as number
      : 0;
    this.state.revision_count += 1;

    const rawDecision = String(criticResult['decision'] ?? 'REVISE');
    this.state.feedback = String(criticResult['actionable_feedback'] ?? '');

    // Record history
    this.state.history.push({
      revision: this.state.revision_count,
      score: this.state.current_score,
      decision: rawDecision,
      feedback_preview: this.state.feedback ? this.state.feedback.substring(0, 100) : '',
    });

    // Check score improvement
    const scoreImprovement = this.state.current_score - this.state.previous_score;
    if (this.state.revision_count > 1 && scoreImprovement < this.config.score_improvement_threshold) {
      this.state.stagnant_count += 1;
    } else {
      this.state.stagnant_count = 0;
    }

    // Determine final decision
    const decision = this._determineDecision(rawDecision, criticResult);
    this.state.decision = decision;

    const roundIdentifier = this._buildRoundIdentifier();
    const artifact = this._buildCheckpointArtifact(roundIdentifier, criticResult);
    this._persistCheckpointArtifact(artifact);
    this.state.feedback_artifacts = this._buildFeedbackArtifacts(roundIdentifier, criticResult);

    return decision;
  }

  private _determineDecision(
    rawDecision: string,
    criticResult: Record<string, unknown>,
  ): RevisionDecision {
    const score = this.state.current_score;

    // Check LOCK analysis C score
    const lockAnalysis = criticResult['lock_analysis'] as Record<string, unknown> | undefined;
    let cScore = 10;  // Default full score
    if (lockAnalysis && typeof lockAnalysis === 'object') {
      const cData = lockAnalysis['C'] as Record<string, unknown> | undefined;
      if (cData && typeof cData === 'object') {
        cScore = typeof cData['score'] === 'number' ? cData['score'] : 10;
      }
    }

    // Decision logic
    if (rawDecision === 'APPROVED') {
      if (score >= this.config.pass_score && cScore >= this.config.min_c_score) {
        return RevisionDecision.APPROVED;
      }
      return RevisionDecision.REVISE;
    }

    if (rawDecision === 'REWRITE') {
      if (this.state.revision_count >= this.config.max_revisions) {
        return RevisionDecision.HUMAN_REVIEW;
      }
      return RevisionDecision.REWRITE;
    }

    if (rawDecision === 'HUMAN_REVIEW') {
      return RevisionDecision.HUMAN_REVIEW;
    }

    // REVISE case
    if (this.state.revision_count >= this.config.max_revisions) {
      return RevisionDecision.HUMAN_REVIEW;
    }

    if (this.state.stagnant_count >= 2) {
      return RevisionDecision.HUMAN_REVIEW;
    }

    return RevisionDecision.REVISE;
  }

  // -----------------------------------------------
  // Feedback / summary accessors
  // -----------------------------------------------

  getFeedbackForWriter(): Record<string, unknown> {
    return {
      feedback: this.state.feedback,
      revision_count: this.state.revision_count,
      previous_score: this.state.previous_score,
      current_score: this.state.current_score,
      history: this.state.history,
      last_checkpoint_id: this.state.last_checkpoint_id,
      checkpoint_trace: this.state.checkpoint_trace,
      quality_mode: this.state.quality_mode,
      requested_quality_level: this.state.requested_quality_level,
      effective_quality_level: this.state.effective_quality_level,
      degrade_reason: this.state.degrade_reason,
      degrade_steps: this.state.degrade_steps,
      feedback_artifacts: this.state.feedback_artifacts,
    };
  }

  getSummary(): Record<string, unknown> {
    return {
      total_revisions: this.state.revision_count,
      final_score: this.state.current_score,
      final_decision: this.state.decision,
      score_trend: (this.state.history as Array<Record<string, unknown>>).map(h => h.score),
      stagnant_count: this.state.stagnant_count,
      history: this.state.history,
      last_checkpoint_id: this.state.last_checkpoint_id,
      checkpoint_trace: this.state.checkpoint_trace,
      quality_mode: this.state.quality_mode,
      requested_quality_level: this.state.requested_quality_level,
      effective_quality_level: this.state.effective_quality_level,
      degrade_reason: this.state.degrade_reason,
      degrade_steps: this.state.degrade_steps,
      feedback_artifacts: this.state.feedback_artifacts,
    };
  }

  reset(): void {
    const normalizedLevel = RevisionLoop.normalizeQualityLevel(this.config.quality_level);
    this.state = createRevisionState(
      this.config.quality_mode,
      normalizedLevel,
      normalizedLevel,
    );
  }
}

// ============================================================
// Feedback artifact envelope builder
// ============================================================

export function buildFeedbackArtifactEnvelope(
  feedbackArtifacts: Array<Record<string, unknown>>,
  sessionId: string,
  runId: string,
  revisionId: string,
  evidenceLinks: string[],
): Record<string, unknown> {
  return {
    artifact_type: 'quality_feedback',
    schema_version: 'evidence.v1',
    date: new Date().toISOString().split('T')[0],
    owner: 'revision_loop',
    input: {
      session_id: sessionId,
      run_id: runId,
      revision_id: revisionId,
    },
    output: {
      feedback_artifacts: feedbackArtifacts,
      count: feedbackArtifacts.length,
    },
    result: 'PASS',
    evidence_links: evidenceLinks,
    trace: {
      session_id: sessionId,
      run_id: runId,
      revision_id: revisionId,
    },
  };
}

// ============================================================
// Session manager (real service from session/session-manager.ts)
// ============================================================

export type ContentType = SessionContentType;

export interface ISessionManager {
  readonly activePath: string;
  readonly archivedPath: string;
  init(options: { sessionId: string; sessionType: string; projectName: string; domain: string }): void;
  write(options: { sessionId: string; contentType: ContentType; content: string; id: string }): void;
  resolvePath(options: { sessionId: string; contentType: ContentType; id: string }): string;
}

// ============================================================
// runRevisionLoop - Top-level async loop
// ============================================================

export type WriterFn = (draft: string, feedback: Record<string, unknown>) => Promise<string>;
export type CriticFn = (draft: string, sceneCard: Record<string, unknown>) => Promise<Record<string, unknown>>;

export interface RunRevisionLoopOptions {
  draft: string;
  sceneCard: Record<string, unknown>;
  writerFn: WriterFn;
  criticFn: CriticFn;
  config?: Partial<RevisionConfig>;
  verbose?: boolean;
  checkpointStore?: Record<string, unknown>;
  checkpointBasePath?: string;
  sessionId?: string;
  sessionManager?: ISessionManager;
}

export async function runRevisionLoop(options: RunRevisionLoopOptions): Promise<Record<string, unknown>> {
  const {
    draft,
    sceneCard,
    writerFn,
    criticFn,
    verbose = true,
    checkpointStore,
    checkpointBasePath,
    sessionId,
    sessionManager,
  } = options;

  const loop = new RevisionLoop(options.config, checkpointStore);
  let currentDraft = draft;

  // Session init (mirrors Python behavior)
  if (sessionManager && sessionId) {
    const activePath = `${sessionManager.activePath}/${sessionId}`;
    const archivedPath = `${sessionManager.archivedPath}/${sessionId}`;
    // In Node.js we skip the filesystem exists check here; the session manager
    // implementation should handle idempotent init.
  }

  while (true) {
    // Evaluate current draft
    if (verbose) {
      _log.info(`\n\uD83E\uDDD0 \u7B2C ${loop.state.revision_count + 1} \u6B21\u8BC4\u4F30...`);
    }

    try {
      const timeoutSeconds = Math.max(loop.config.quality_phase_timeout_seconds, 0);
      let criticResult: Record<string, unknown>;

      if (timeoutSeconds > 0) {
        criticResult = await Promise.race([
          criticFn(currentDraft, sceneCard),
          new Promise<never>((_, reject) =>
            setTimeout(() => reject(new Error('TimeoutError')), timeoutSeconds * 1000),
          ),
        ]);
      } else {
        criticResult = await criticFn(currentDraft, sceneCard);
      }

      if (sessionId && typeof criticResult === 'object' && !criticResult['session_id']) {
        criticResult = { ...criticResult, session_id: sessionId };
      }

      const decision = loop.updateFromCritic(criticResult);

      if (verbose) {
        _log.info(`   \u5206\u6570: ${loop.state.current_score.toFixed(1)}`);
        _log.info(`   \u51B3\u7B56: ${decision}`);
      }

      // Check whether to continue
      if (!loop.shouldContinue()) break;

      // Get feedback and revise
      const feedback = loop.getFeedbackForWriter();

      if (verbose) {
        _log.info(`\n\u270D\uFE0F \u7B2C ${loop.state.revision_count} \u6B21\u4FEE\u8BA2...`);
      }

      try {
        const writerTimeout = Math.max(loop.config.quality_phase_timeout_seconds, 0);
        if (writerTimeout > 0) {
          currentDraft = await Promise.race([
            writerFn(currentDraft, feedback),
            new Promise<never>((_, reject) =>
              setTimeout(() => reject(new Error('TimeoutError')), writerTimeout * 1000),
            ),
          ]);
        } else {
          currentDraft = await writerFn(currentDraft, feedback);
        }
      } catch (exc) {
        const errorMsg = exc instanceof Error ? exc.message : String(exc);
        const isTimeout = errorMsg === 'TimeoutError';

        if (isTimeout) {
          const degraded = loop.handleRuntimeEvent('timeout', 'writer');
          if (degraded) {
            if (verbose) {
              _log.info(`   \u8D28\u91CF\u964D\u7EA7: writer timeout -> ${loop.state.effective_quality_level}`);
            }
            continue;
          }
          loop.state.decision = RevisionDecision.HUMAN_REVIEW;
          loop.state.degrade_reason = 'timeout:writer';
          break;
        } else {
          const detail = exc instanceof Error ? exc.constructor.name : 'UnknownError';
          const degraded = loop.handleRuntimeEvent('error', 'writer', detail);
          if (degraded) {
            if (verbose) {
              _log.info(`   \u8D28\u91CF\u964D\u7EA7: writer error -> ${loop.state.effective_quality_level}`);
            }
            continue;
          }
          loop.state.decision = RevisionDecision.HUMAN_REVIEW;
          loop.state.degrade_reason = `error:writer:${detail}`;
          break;
        }
      }
    } catch (exc) {
      const errorMsg = exc instanceof Error ? exc.message : String(exc);
      const isTimeout = errorMsg === 'TimeoutError';

      if (isTimeout) {
        const degraded = loop.handleRuntimeEvent('timeout', 'critic');
        if (degraded) {
          if (verbose) {
            _log.info(`   \u8D28\u91CF\u964D\u7EA7: critic timeout -> ${loop.state.effective_quality_level}`);
          }
          continue;
        }
        loop.state.decision = RevisionDecision.HUMAN_REVIEW;
        loop.state.degrade_reason = 'timeout:critic';
        break;
      } else {
        const detail = exc instanceof Error ? exc.constructor.name : 'UnknownError';
        const degraded = loop.handleRuntimeEvent('error', 'critic', detail);
        if (degraded) {
          if (verbose) {
            _log.info(`   \u8D28\u91CF\u964D\u7EA7: critic error -> ${loop.state.effective_quality_level}`);
          }
          continue;
        }
        loop.state.decision = RevisionDecision.HUMAN_REVIEW;
        loop.state.degrade_reason = `error:critic:${detail}`;
        break;
      }
    }
  }

  // Return result
  const summary = loop.getSummary();
  (summary as Record<string, unknown>)['final_draft'] = currentDraft;

  if (verbose) {
    _log.info(`\n${'='.repeat(50)}`);
    _log.info('\u4FEE\u8BA2\u5FAA\u73AF\u5B8C\u6210');
    _log.info(`   \u603B\u4FEE\u8BA2\u6B21\u6570: ${summary.total_revisions}`);
    _log.info(`   \u6700\u7EC8\u5206\u6570: ${(summary.final_score as number).toFixed(1)}`);
    _log.info(`   \u6700\u7EC8\u51B3\u7B56: ${summary.final_decision}`);
    _log.info('='.repeat(50));
  }

  return summary;
}
